pkg/k8s/resource/resource.go

```diff
@@ -0,0 +1,307 @@
+		maxRetries:            r.maxEventRetries,
+	}
+
+	// Fork a goroutine to pop elements from the queue and pass them to the subscriber.
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		defer subCancel()
+
+		// Wait for the store so we can emit the initial items and the sync event.
+		store := r.storePromise.Await()
+
+		r.mu.Lock()
+		r.subId++
+		r.queues[r.subId] = queue
+
+		// Append the initial set of keys to the queue + sentinel for the sync event.
+		// We go through the queue instead of directly emitting to avoid a race between
+		// the queue add and listing keys, plus to have one error handling path for Event.Done().
+		keyIter := store.IterKeys()
+		for keyIter.Next() {
+			fmt.Printf("adding %q\n", keyIter.Key())
+			queue.Add(keyIter.Key())
+		}
+		queue.Add(syncSentinelKey)
+		r.mu.Unlock()
+
+		for {
+			// Retrieve an item from the subscribers queue and then fetch the object
+			// from the store.
+			rawKey, shutdown := queue.Get()
+			if shutdown {
+				break
+			}
+			key := rawKey.(Key)
+			base := baseEvent{queue.eventDone, key}
+
+			if key == syncSentinelKey {
+				next(&SyncEvent[T]{base, store})
+			} else {
+				obj, exists, err := store.GetByKey(key)
+				if err != nil {
+					queue.setError(err)
+					break
+				}
+
+				if exists {
+					next(&UpdateEvent[T]{base, obj})
+				} else {
+					next(&DeleteEvent[T]{base})
+				}
+			}
+		}
+
+		r.mu.Lock()
+		delete(r.queues, r.subId)
+		r.mu.Unlock()
+		complete(queue.getError())
+	}()
+
+	// Fork a goroutine to wait for either the subscriber cancelling or the resource
+	// shutting down.
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		select {
+		case <-r.ctx.Done():
+			subCancel()
+		case <-subCtx.Done():
+		}
+		queue.ShutDownWithDrain()
+	}()
+}
+
+// keyQueue wraps the workqueue to implement the error retry logic for a single subscriber,
+// e.g. it implements the eventDone() method called by Event[T].Done().
+type keyQueue struct {
+	lock.Mutex
+	workqueue.RateLimitingInterface
+
+	maxRetries int
+	err        error
+}
+
+func (kq *keyQueue) getError() error {
+	kq.Lock()
+	defer kq.Unlock()
+	return kq.err
+}
+
+func (kq *keyQueue) setError(err error) {
+	kq.Lock()
+	defer kq.Unlock()
+	if kq.err == nil {
+		kq.err = err
+	}
+}
+
+func (kq *keyQueue) eventDone(key Key, err error) {
+	// This is based on the example found in k8s.io/client-go/examples/workqueue/main.go.
+
+	// Mark the object as done being processed. If it was marked dirty
+	// during processing, it'll be processed again.
+	defer kq.Done(key)
+
+	if err != nil {
+		// Processing of this item failed. Check if retry limit has been reached.
+		if kq.NumRequeues(key) >= kq.maxRetries-1 {
+			// TODO: check that it's safe to call ShutDown here (no deadlock)
+			kq.setError(err)
+			kq.ShutDown()
+			return
+		}
+
+		// Can still retry processing it. Add it back to the queue.
+		// Blocks until rate limiter says it is OK to continue.
+		kq.AddRateLimited(key)
```

Comment:
TODO: This should probably be called in the background as we don't want to block other work from being processed?